driver/src/orderbook/streamed/block_timestamp.rs

```diff
@@ -21,3 +21,36 @@ impl BlockTimestamp for Web3 {
         .boxed()
     }
 }
+
+/// A cache for the block timestamp which avoids having to query the node in the case where we
+/// receive multiple events from the same block in a row.
+#[derive(Debug)]
+pub struct BlockTimestampCache<T> {
+    block_timestamp: T,
+    hash: H256,
+    timestamp: u64,
+}
+
+impl<T> BlockTimestampCache<T> {
+    pub fn new(block_timestamp: T) -> Self {
+        Self {
+            block_timestamp,
+            hash: H256::zero(),
+            timestamp: 0,
+        }
+    }
+}
+
+impl<T: BlockTimestamp + Send> BlockTimestamp for BlockTimestampCache<T> {
```

Comment:
This is so cool! Layering abstractions! :tada:

Reply:
Is there value in having no type specialization in the struct definition (and only adding it here). I think it hides a bit the purpose of this struct behind unused generalization.

Reply:
I prefer requiring bounds only in the place where needed. That makes it more clear to me why they are required.